pkg/icingadb/db.go

```diff
@@ -296,6 +311,49 @@ func (db *DB) NamedBulkExec(
 	return g.Wait()
 }
 
+// NamedExecContextBulk performs db.NamedExecContext(ctx, query, rows).
+// If that fails due to upserting the same row twice (Postgres only), it splits rows and upserts them in a transaction.
+func (db *DB) NamedExecContextBulk(ctx context.Context, query string, rows []contracts.Entity) error {
+	_, err := db.NamedExecContext(ctx, query, rows)
+	if err != nil && len(rows) > 1 && isDuplicateRowInUpsert(err) {
+		tx, err := db.BeginTxx(ctx, nil)
+		if err != nil {
+			return errors.Wrap(err, "can't start transaction")
+		}
+
+		chunk := rows[:len(rows)/2]
+
+		for {
+			if _, err := tx.NamedExecContext(ctx, query, chunk); err != nil {
+				if len(chunk) > 1 && isDuplicateRowInUpsert(err) {
+					chunk = chunk[:len(chunk)/2]
+					continue
+				}
+
+				return errors.Wrap(err, "can't execute statement in transaction")
+			}
+
+			if chunk = rows[len(chunk):]; len(chunk) > 0 {
+				continue
+			}
+
+			break
+		}
+
+		if err := tx.Commit(); err != nil {
+			return errors.Wrap(err, "can't commit transaction")
+		}
+	}
+
+	return err
+}
```

Comment:
In what kind of situations does this happen? Ideally we shouldn't get into this situation at all as updating the same row twice is a waste of time anyways.

Reply:
History.

Reply:
Do you actually need both updates there or could you just omit the first one?

Reply:
They could affect different columns, so yes.

Reply:
I think it would be nicer if we could never send upserts with duplicate keys to Postgres in the first place (this would also avoid having to perform string matching on error messages). The situation should happen if and only if there are multiple entities in `rows` that return the same value for `rows[i].ID()`. So I think it should be possible to just check this before sending the first query and then splitting the rows into a new query before each duplicate.